check for max dimensions

Jaapel opened the pull request:
Fixes #111 .
Tests use `unittest.mock.MagicMock` to avoid creating large arrays during testing.

Jaapel committed:
check for max dimensions

Jaapel commented:
@mirzaees let me know what you think!

scottstanie commented on src/snaphu/_check.py:
```diff
@@ -70,6 +73,71 @@ def check_dataset_shapes(
+    elif isinstance(tile_overlap, tuple):
+        # cannot unpack, tile_overlap shape is checked later
+        x_overlap = tile_overlap[0]
+        y_overlap = tile_overlap[1]
```
```suggestion
    if isinstance(tile_overlap, int):
        y_overlap = x_overlap = tile_overlap
    elif isinstance(tile_overlap, tuple):
        # cannot unpack, tile_overlap shape is checked later
        y_overlap = tile_overlap[0]
        x_overlap = tile_overlap[1]
```

scottstanie replied:
(though i'm seeing that elsewhere in the code it a little ambiguous if the tuple is (row, col) or (x, y). i'm the tuple is `(row, column)` )

Jaapel replied:
Fixed this

Jaapel replied:
The `ntiles` docs say row / col, so I may have to revert this: https://github.com/Jaapel/snaphu-py/blob/afe62434dfeddf9875dadf09784c412228f38d5c/src/snaphu/_unwrap.py#L213. Can you confirm?

scottstanie commented on src/snaphu/_check.py:
```diff
@@ -70,6 +73,71 @@ def check_dataset_shapes(
+        tile_shapes_max = (
+            arr.shape[0] + 1 // ntiles[0] + x_overlap,
+            arr.shape[1] + 1 // ntiles[1] + y_overlap,
+        )
```
was this logic taken from somewhere in snaphu? i dont think it looks right... ` + 1 // ntiles[0]` will always either be 1 or 0, and this would . maybe you want something like 

```suggestion
        # in case tile exceeds max array size
        tile_height = math.ceil(arr.shape[0] / ntiles[0]) + y_overlap
        tile_width  = math.ceil(arr.shape[1] / ntiles[1]) + x_overlap
        tile_shapes_max = (tile_height, tile_width)

```
but it could be good to check the C code for the actual max shape

Jaapel replied:
I have checked the C code now

Jaapel replied:
I forget the `(` in the answer you've reviewed. I also updated the tests to be more precise.

Jaapel committed:
match calc snaphu.c

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -15,6 +17,9 @@
 ]
 
 
+LARGESHORT = 32000
```
Let's add a comment here to remind us that this constant needs to match the corresponding one in the SNAPHU C code.

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -70,6 +75,74 @@ def check_dataset_shapes(
+    **datasets: InputDataset | OutputDataset,
+) -> None:
+    """
+    Ensure that one or more datasets have shape that snaphu can handle.
```
Let's always capitalize "SNAPHU" since that's how it's stylized on the [official website](https://web.stanford.edu/group/radar/softwareandlinks/sw/snaphu/) and in journal papers.

```suggestion
    Ensure that one or more datasets have shape that SNAPHU can handle.
```

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -70,6 +75,74 @@ def check_dataset_shapes(
+    single_tile_reoptimize: bool
+        Whether to use single tile reoptimization after tiled unwrapping.
+    **datasets : dict, optional
+        Datasets whose shape must be equal to `shape`. The name of each keyword argument
```
`shape` isn't an argument here so let's update this docstring description.

```suggestion
        Datasets to be processed with SNAPHU. The name of each keyword argument
```

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -70,6 +75,74 @@ def check_dataset_shapes(
+        # cannot unpack, tile_overlap shape is checked later
+        y_overlap = tile_overlap[0]
+        x_overlap = tile_overlap[1]
+
+    else:
```
Remove this blank line please.

```suggestion
        x_overlap = tile_overlap[1]
    else:
```

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -70,6 +75,74 @@ def check_dataset_shapes(
+        x_overlap = tile_overlap[1]
+
+    else:
+        msg = f"Unknown format tile overlaps: {tile_overlap}"
```
Let's update this error message so that it mentions the type of `tile_overlap` in order to help with debugging.

```suggestion
        msg = f"got {type(tile_overlap)=}, expected int or tuple"
```

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -70,6 +75,74 @@ def check_dataset_shapes(
+        raise TypeError(msg)
+
+    for name, arr in datasets.items():
+        skip_tiling: bool = ntiles == (1, 1)
```
This line doesn't depend on `name` or `arr` so let's pull it out of the loop.

Do we really need the type hint here? I assume that mypy can infer that `skip_tiling` is a `bool` on its own.

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -70,6 +75,74 @@ def check_dataset_shapes(
+            for size in arr.shape:
+                if size > LARGESHORT:
+                    msg = f"dataset {name} too large for snaphu, shape: {arr.shape}"
+                    raise ValueError(msg)
```
Let's refactor this slightly to remove one level of indentation.

```suggestion
            if any(n > LARGESHORT for n in arr.shape):
                msg = f"dataset {name} too large for SNAPHU, shape: {arr.shape}"
                raise ValueError(msg)
```

(Same for the similar check below)

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -70,6 +75,74 @@ def check_dataset_shapes(
+                    msg = f"dataset {name} too large for snaphu, shape: {arr.shape}"
+                    raise ValueError(msg)
+            if skip_tiling:
+                return
```
Is it correct to `return` here? This could cause the function to exit during the first iteration of the loop without checking the size of each subsequent dataset.

```suggestion
                continue
```

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -70,6 +75,74 @@ def check_dataset_shapes(
+            # a single tile is input for snaphu
+            for size in arr.shape:
+                if size > LARGESHORT:
+                    msg = f"dataset {name} too large for snaphu, shape: {arr.shape}"
```
Let's add some suggestions here to help users work around this.

```suggestion
                    msg = (
                        f"{name} dataset with shape {arr.shape} exceeds max dimensions"
                        " supported by SNAPHU. Consider using tiling and disabling the"
                        " regrow_conncomps and single_tile_reoptimize options"
                    )
```

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -70,6 +75,74 @@ def check_dataset_shapes(
+                return
+
+        # in case tile exceeds max array size
+        def _calc_tile_shape(array_len: int, num_tiles: int, overlap_len: int):
```
Let's annotate the return type here as well.

```suggestion
        def _calc_tile_shape(array_len: int, num_tiles: int, overlap_len: int) -> int:
```

gmgunter commented on src/snaphu/_check.py:
```diff
@@ -70,6 +75,74 @@ def check_dataset_shapes(
+        tile_shapes_max = (tile_height, tile_width)
+        for size in tile_shapes_max:
+            if size > LARGESHORT:
+                msg = f"dataset {name} too large for snaphu, shape: {arr.shape}"
```
Let's update this error message for clarity. `arr.shape` isn't directly relevant here -- it's the tile dimensions that are too large. Maybe something like this:

```suggestion
                msg = (
                    f"tile dimensions for {name} dataset are {tile_shapes_max}, which"
                    " exceed the max supported by SNAPHU. Consider increasing number"
                    " of tiles"
                )
```

gmgunter commented on src/snaphu/_unwrap.py:
```diff
@@ -332,6 +333,24 @@ def unwrap(  # type: ignore[no-untyped-def]
+            regrow_conncomps=regrow_conncomps,
+            single_tile_reoptimize=single_tile_reoptimize,
+            mask=mask,
+        )
```
What about the `igram` and `corr` input datasets? Should we check their sizes as well?

There's an earlier [check](https://github.com/isce-framework/snaphu-py/blob/afe62434dfeddf9875dadf09784c412228f38d5c/src/snaphu/_unwrap.py#L304-L307) that ensures that all of the input & output arrays have the same shape, so actually I think it's sufficient to just check `igram`'s shape.

```suggestion
    # Ensure that the dataset (and tile) dimensions are not too large.
    check_dataset_sizes(
        ntiles,
        tile_overlap,
        regrow_conncomps=regrow_conncomps,
        single_tile_reoptimize=single_tile_reoptimize,
        igram=igram,
    )
```

gmgunter reviewed (commented):
Thanks for taking the time to improve the codebase. Nice use of mocking in the unit test.

Jaapel committed:
Capitalize snaphu in  src/snaphu/_check.py

Co-authored-by: Geoffrey Gunter <12984092+gmgunter@users.noreply.github.com>

Jaapel committed:
Update datasets docs in src/snaphu/_check.py

Co-authored-by: Geoffrey Gunter <12984092+gmgunter@users.noreply.github.com>

Jaapel committed:
remove blank line in src/snaphu/_check.py

Co-authored-by: Geoffrey Gunter <12984092+gmgunter@users.noreply.github.com>

Jaapel committed:
better specify type error in src/snaphu/_check.py

Co-authored-by: Geoffrey Gunter <12984092+gmgunter@users.noreply.github.com>

Jaapel committed:
simplify looping logic in src/snaphu/_check.py

Co-authored-by: Geoffrey Gunter <12984092+gmgunter@users.noreply.github.com>

Jaapel committed:
fix mistake in looping through datsets in src/snaphu/_check.py

Co-authored-by: Geoffrey Gunter <12984092+gmgunter@users.noreply.github.com>

Jaapel committed:
typehint _calc_tile_shape in src/snaphu/_check.py

Co-authored-by: Geoffrey Gunter <12984092+gmgunter@users.noreply.github.com>

Jaapel committed:
Update error message on tile dimensions in src/snaphu/_check.py

Co-authored-by: Geoffrey Gunter <12984092+gmgunter@users.noreply.github.com>

Jaapel committed:
only need to check size of igram in src/snaphu/_unwrap.py

Co-authored-by: Geoffrey Gunter <12984092+gmgunter@users.noreply.github.com>

Jaapel committed:
update error messages

Jaapel commented:
Thanks for the extended review! I implemented all your suggestions @gmgunter .
@scottstanie I implemented your changes, but based on the `unwrap` docs, `ntiles` is a (row, col) tuple. So I guess I need to rename `y` to row and `x` to col?

Jaapel committed:
use backwards compatible dtypes

mirzaees commented:
@scottstanie and @gmgunter please approve this PR if it is final
